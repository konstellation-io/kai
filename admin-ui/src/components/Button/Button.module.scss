@import '../../styles/colors';
@import '../../styles/mixins';
@import '../../styles/variables';

.btn {
  @include font-button;
  position: relative;
  text-align: center;
  padding: 0 42px;
  white-space: nowrap;
  cursor: pointer;
  transition: color ease 0.4s, background-color ease 0.4s;

  display: flex;
  flex-direction: row;
  align-items: center;

  span {
    width: 100%;
  }

  svg {
    margin-right: 14%;
  }

  span {
    width: 100%;
  }

  &.middle {
    text-align: middle;
    padding: 0 20px;
  }
  &.left {
    text-align: left;
    padding: 0 20px;
  }
  &.right {
    text-align: right;
    padding: 0 20px;
  }

  &:hover {
    color: $font-color-white;
  }

  &:active {
    color: $font-color;
  }

  &.border {
    border: solid 1px $color-blue;
  }
}

.primary {
  color: #1a1b1f;
  background-color: $color-button-blue;

  &:active {
    box-shadow: 0 0 20px 0 $color-button-blue;
  }
}

.dark {
  color: $font-color;
  background-color: black;

  &:hover {
    color: $font-color-white;
  }

  &:active {
    box-shadow: none;
  }
}

.grey {
  color: $font-color;
  background-color: $color-grey-darker;

  &:hover {
    color: $font-color-white;
  }

  &:active {
    box-shadow: none;
  }
}

.transparent {
  color: $font-color;
  background-color: transparent;

  &:hover {
    color: $font-color-white;
  }

  &:active {
    box-shadow: none;
  }

  &.primary {
    background-color: rgba(191, 207, 255, 0.15);
    color: white;
  }
}

.label {
  color: $font-color;
}

.disabled {
  color: $font-color;
  pointer-events: none;
  background-color: $color-grey;
}
