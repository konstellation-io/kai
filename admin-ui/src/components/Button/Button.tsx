import React from 'react';
import { Link } from 'react-router-dom';

import Spinner from '../../components/Spinner/Spinner';

import cx from 'classnames';
import styles from './Button.module.scss';

export const BUTTON_TYPES = {
  DEFAULT: 'default',
  DARK: 'dark',
  GREY: 'grey',
  TRANSPARENT: 'transparent'
};
export const BUTTON_ALIGN = {
  LEFT: 'left',
  MIDDLE: 'middle',
  RIGHT: 'right'
};

type Props = {
  label?: string;
  type?: string;
  border?: boolean;
  Icon?: any;
  to?: string;
  onClick?: any;
  primary?: boolean;
  disabled?: boolean;
  loading?: boolean;
  height?: number;
  align?: string;
  style?: Object;
};

function Button({
  type = BUTTON_TYPES.DEFAULT,
  border = false,
  label = 'Button',
  Icon = undefined,
  to = '',
  onClick = function() {},
  primary = false,
  disabled = false,
  loading = false,
  height = 40,
  align = BUTTON_ALIGN.MIDDLE,
  style = {}
}: Props) {
  const content = loading ? (
    <Spinner size={30} color="#0d0e11" />
  ) : (
    <>
      {Icon && <Icon style={{ fontSize: '1rem' }} />}
      <span>{label}</span>
    </>
  );

  const btn = (
    <div
      className={cx(styles.btn, styles[type], styles[align], {
        [styles.primary]: primary,
        [styles.border]: border,
        [styles.label]: !primary,
        [styles.disabled]: disabled
      })}
      style={{ ...style, height, lineHeight: `${height}px` }}
      onClick={onClick}
    >
      {content}
    </div>
  );

  return !disabled && to !== '' ? (
    <Link to={to} data-testid="buttonLink">
      {btn}
    </Link>
  ) : (
    btn
  );
}

export default Button;
