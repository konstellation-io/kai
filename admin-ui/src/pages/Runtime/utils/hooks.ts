import { useMutation } from '@apollo/react-hooks';
import {
  ACTIVATE_VERSION,
  DEACTIVATE_VERSION,
  DEPLOY_VERSION,
  STOP_VERSION,
  ActivateVersionResponse,
  DeactivateVersionResponse,
  DeployVersionResponse,
  StopVersionResponse,
  VersionActionVars
} from '../pages/RuntimeStatusPreview/RuntimeStatusPreview.graphql';

export enum versionActions {
  activate = 'activateVersion',
  deploy = 'deployVersion',
  stop = 'stopVersion',
  deactivate = 'deactivateVersion'
}

export default function useVersionAction() {
  const [activateMutation, { loading: loadingM1 }] = useMutation<
    ActivateVersionResponse,
    VersionActionVars
  >(ACTIVATE_VERSION);
  const [deactivateMutation, { loading: loadingM2 }] = useMutation<
    DeactivateVersionResponse,
    VersionActionVars
  >(DEACTIVATE_VERSION);
  const [deployMutation, { loading: loadingM3 }] = useMutation<
    DeployVersionResponse,
    VersionActionVars
  >(DEPLOY_VERSION);
  const [stopMutation, { loading: loadingM4 }] = useMutation<
    StopVersionResponse,
    VersionActionVars
  >(STOP_VERSION);

  const mutationLoading = [loadingM1, loadingM2, loadingM3, loadingM4].some(
    el => el
  );

  function getMutationVars(versionId: string, comment?: string) {
    const variables = {
      variables: {
        input: {
          versionId: versionId
        }
      }
    } as any;

    if (comment) {
      variables.variables.input.comment = comment;
    }

    return variables;
  }

  return {
    [versionActions.activate]: activateMutation,
    [versionActions.deploy]: deployMutation,
    [versionActions.stop]: stopMutation,
    [versionActions.deactivate]: deactivateMutation,
    getMutationVars,
    mutationLoading
  };
}
