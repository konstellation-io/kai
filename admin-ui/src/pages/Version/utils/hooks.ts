import { loader } from 'graphql.macro';
import { useMutation } from '@apollo/react-hooks';
import {
  GetVersionConfStatus,
  GetVersionConfStatusVariables
} from '../../../graphql/queries/types/GetVersionConfStatus';
import {
  StartVersion,
  StartVersionVariables
} from '../../../graphql/mutations/types/StartVersion';
import {
  StopVersion,
  StopVersionVariables
} from '../../../graphql/mutations/types/StopVersion';
import {
  UnpublishVersion,
  UnpublishVersionVariables
} from '../../../graphql/mutations/types/UnpublishVersion';
import {
  PublishVersion,
  PublishVersionVariables
} from '../../../graphql/mutations/types/PublishVersion';
import { VersionStatus } from '../../../graphql/types/globalTypes';

const PublishVersionMutation = loader(
  '../../../graphql/mutations/publishVersion.graphql'
);
const UnpublishVersionMutation = loader(
  '../../../graphql/mutations/unpublishVersion.graphql'
);
const StartVersionMutation = loader(
  '../../../graphql/mutations/startVersion.graphql'
);
const StopVersionMutation = loader(
  '../../../graphql/mutations/stopVersion.graphql'
);
const GetRuntimeAndVersionsQuery = loader(
  '../../../graphql/queries/getRuntimeAndVersions.graphql'
);

export enum versionActions {
  start = 'startVersion',
  stop = 'stopVersion',
  publish = 'publishVersion',
  unpublish = 'unpublishVersion'
}

export default function useVersionAction(runtimeId: string) {
  const [publishMutation, { loading: loadingM1 }] = useMutation<
    PublishVersion,
    PublishVersionVariables
  >(PublishVersionMutation, {
    update(cache, updateResult) {
      // TODO: This update the previous one activated version.
      // We should remove this when multi activation feature is implemented.
      if (updateResult.data !== undefined && updateResult.data !== null) {
        const updatedVersion = updateResult.data.publishVersion;

        const cacheResult = cache.readQuery<
          GetVersionConfStatus,
          GetVersionConfStatusVariables
        >({
          query: GetRuntimeAndVersionsQuery,
          variables: {
            runtimeId: runtimeId
          }
        });

        if (cacheResult !== null) {
          const { versions, runtime } = cacheResult;
          const newVersions = versions.map(v => {
            if (v.id === updatedVersion.id) return v;
            if (v.status === VersionStatus.PUBLISHED) {
              v.status = VersionStatus.STARTED;
            }
            return v;
          });

          cache.writeQuery({
            query: GetRuntimeAndVersionsQuery,
            data: {
              runtime,
              versions: newVersions
            }
          });
        }
      }
    }
  });

  const [unpublishMutation, { loading: loadingM2 }] = useMutation<
    UnpublishVersion,
    UnpublishVersionVariables
  >(UnpublishVersionMutation);
  const [startMutation, { loading: loadingM3 }] = useMutation<
    StartVersion,
    StartVersionVariables
  >(StartVersionMutation);
  const [stopMutation, { loading: loadingM4 }] = useMutation<
    StopVersion,
    StopVersionVariables
  >(StopVersionMutation);

  const mutationLoading = [loadingM1, loadingM2, loadingM3, loadingM4].some(
    el => el
  );

  type MutationsVariables = {
    variables: PublishVersionVariables;
  };

  function getMutationVars(
    versionId: string,
    comment: string
  ): MutationsVariables {
    return { variables: { input: { versionId, comment } } };
  }

  return {
    [versionActions.publish]: publishMutation,
    [versionActions.start]: startMutation,
    [versionActions.stop]: stopMutation,
    [versionActions.unpublish]: unpublishMutation,
    getMutationVars,
    mutationLoading
  };
}
