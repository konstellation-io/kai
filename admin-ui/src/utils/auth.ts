import Cookies from 'js-cookie';

export const JWT_COOKIE = '__jwt';

export const isUserAuthenticated = () => {
  const jwt = Cookies.get(JWT_COOKIE);

  return jwt;
};
