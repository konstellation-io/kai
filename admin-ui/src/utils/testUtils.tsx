import React from 'react';
import { mount } from 'enzyme';

type TestHookprops = {
  callback: Function;
};
const TestHook = ({ callback }: TestHookprops) => {
  callback();
  return null;
};

export const testHook = (callback: Function) => {
  mount(<TestHook callback={callback} />);
};
