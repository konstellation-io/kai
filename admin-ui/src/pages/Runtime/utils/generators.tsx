import React from 'react';
import Button from '../../../components/Button/Button';
import StartIcon from '@material-ui/icons/SkipNext';
import StopIcon from '@material-ui/icons/PauseCircleFilled';
import PublishIcon from '@material-ui/icons/PlayCircleFilledOutlined';
import UnpublishIcon from '@material-ui/icons/Block';
import { VersionStatus } from '../../../graphql/types/globalTypes';

function generateActionButton(label: string, icon: any, action: Function) {
  return (
    <Button
      key={label}
      label={label}
      onClick={action}
      height={30}
      Icon={icon}
    />
  );
}

function getStateToButtons(
  publishAction: Function,
  startAction: Function,
  stopAction: Function,
  unpublishAction: Function
) {
  const buttonStart = generateActionButton('START', StartIcon, startAction);
  const buttonStop = generateActionButton('STOP', StopIcon, stopAction);
  const buttonPublish = generateActionButton(
    'PUBLISH',
    PublishIcon,
    publishAction
  );
  const buttonUnpublish = generateActionButton(
    'UNPUBLISH',
    UnpublishIcon,
    unpublishAction
  );

  return {
    [VersionStatus.STOPPED]: [buttonStart],
    [VersionStatus.PUBLISHED]: [buttonUnpublish],
    [VersionStatus.STARTED]: [buttonPublish, buttonStop],
    [VersionStatus.STOPPED]: [buttonStart]
  };
}

export function getVersionActionButtons(
  publishAction: Function,
  startAction: Function,
  stopAction: Function,
  unpublishAction: Function,
  status?: string
) {
  const stateToButtons: { [key: string]: any } = getStateToButtons(
    publishAction,
    startAction,
    stopAction,
    unpublishAction
  );

  return stateToButtons[status || ''];
}
