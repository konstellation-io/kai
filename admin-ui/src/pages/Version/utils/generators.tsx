import React, { ReactElement, MouseEvent, FunctionComponent } from 'react';
import Button from '../../../components/Button/Button';
import StartIcon from '@material-ui/icons/SkipNext';
import StopIcon from '@material-ui/icons/PauseCircleFilled';
import PublishIcon from '@material-ui/icons/PlayCircleFilledOutlined';
import UnpublishIcon from '@material-ui/icons/Block';
import { VersionStatus } from '../../../graphql/types/globalTypes';
import { SvgIconProps } from '@material-ui/core/SvgIcon';

function generateActionButton(
  label: string,
  icon: FunctionComponent<SvgIconProps>,
  action: (e: MouseEvent<HTMLDivElement>) => void
): ReactElement {
  return (
    <Button
      key={label}
      label={label}
      onClick={action}
      height={30}
      Icon={icon}
    />
  );
}

function getStateToButtons(
  publishAction: (e: MouseEvent<HTMLDivElement>) => void,
  startAction: (e: MouseEvent<HTMLDivElement>) => void,
  stopAction: (e: MouseEvent<HTMLDivElement>) => void,
  unpublishAction: (e: MouseEvent<HTMLDivElement>) => void
): { [key: string]: ReactElement[] } {
  const buttonStart = generateActionButton('START', StartIcon, startAction);
  const buttonStop = generateActionButton('STOP', StopIcon, stopAction);
  const buttonPublish = generateActionButton(
    'PUBLISH',
    PublishIcon,
    publishAction
  );
  const buttonUnpublish = generateActionButton(
    'UNPUBLISH',
    UnpublishIcon,
    unpublishAction
  );

  return {
    [VersionStatus.STOPPED]: [buttonStart],
    [VersionStatus.PUBLISHED]: [buttonUnpublish],
    [VersionStatus.STARTED]: [buttonPublish, buttonStop],
    [VersionStatus.STOPPED]: [buttonStart]
  };
}

export function getVersionActionButtons(
  publishAction: (e: MouseEvent<HTMLDivElement>) => void,
  startAction: (e: MouseEvent<HTMLDivElement>) => void,
  stopAction: (e: MouseEvent<HTMLDivElement>) => void,
  unpublishAction: (e: MouseEvent<HTMLDivElement>) => void,
  status?: string
) {
  const stateToButtons = getStateToButtons(
    publishAction,
    startAction,
    stopAction,
    unpublishAction
  );

  if (status === undefined) {
    return [];
  }

  return stateToButtons[status];
}
