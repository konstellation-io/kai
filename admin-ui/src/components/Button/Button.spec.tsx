import React from 'react';
import { renderWithRouter } from '../../utils/testUtils';
import { fireEvent, cleanup } from '@testing-library/react';
import { getByTestId } from '@testing-library/dom';
import Button from './Button';

afterEach(cleanup);

it('Renders Button without crashing', () => {
  const {
    element: { container }
  } = renderWithRouter(<Button />);
  expect(container).toMatchSnapshot();
});

it('Shows right texts', () => {
  const {
    element: { customRerender, container }
  } = renderWithRouter(<Button />);
  expect(container.textContent).toBe('Button');

  customRerender(<Button label="New Text" />);
  expect(container.textContent).toBe('New Text');

  customRerender(
    <Button
      label="Other Text"
      type="dark"
      onClick={function() {}}
      primary
      disabled
    />
  );
  expect(container.textContent).toBe('Other Text');
});

it('Handles click events', () => {
  const clickMock = jest.fn(() => true);

  const {
    element: { getByText }
  } = renderWithRouter(<Button onClick={clickMock} />);

  fireEvent.click(getByText('Button'));
  expect(clickMock).toHaveBeenCalledTimes(1);
});

it('shows loader when loading', () => {
  const {
    element: { container }
  } = renderWithRouter(<Button loading />);

  expect(getByTestId(container, 'spinner')).toBeInTheDocument();
});

it('is a route button when a path is given', () => {
  const {
    element: { container }
  } = renderWithRouter(<Button to="/login/:someParam" />);

  expect(getByTestId(container, 'buttonLink')).toBeInTheDocument();
});
