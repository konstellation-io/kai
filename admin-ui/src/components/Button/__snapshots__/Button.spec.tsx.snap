// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Button without crashing 1`] = `
<div>
  <div
    class="btn default middle label"
    style="height: 40px; line-height: 40px;"
  >
    <span>
      Button
    </span>
  </div>
</div>
`;
